### PR TITLE
webhook Handle, patch only kubemacpool changes

### PR DESCRIPTION
**Pull request by RamLavi**
**What this PR does / why we need it**:
Currently kubemacpool webhook handlers patch the entire request object
(pod/virtualMachine). But sometimes there are changes made by other webhooks
or from internal kubernetes changes, that we don't want to set in our webhook.
In order to only add the wanted changes,
we change our webhook patch strategy to only patch the relevant changes.

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by RamLavi** [comment]
/hold

let's merge this change only after we fixed the KMP issue causing test 2995 to be [flaky](https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_cluster-network-addons-operator/689/pull-e2e-cluster-network-addons-operator-kubemacpool-functests/1336970017039716352) on CNAO tier1 tests, as I believe this change may reduce it's appearance.

File: pkg/webhook/pod/pod.go
Comment (RamLavi) [inline pkg/webhook/pod/pod.go:76]: I think I can make a good reuse of `patchChange()` [function](https://github.com/k8snetworkplumbingwg/kubemacpool/pull/259/files#diff-276495201a043c6a12276502c3cf1a1db02fd2c7c9f4323b20415c49b244e9dbR125), but not sure where I should place it...

File: pkg/webhook/pod/pod.go
Comment (phoracek) [inline pkg/webhook/pod/pod.go:72]: Since we know all the fields we intent to control, making diff of the whole objects seems redundant. Not only that, onl line 66 we change the pod in-place. AllocatePodMac is a black box and we just assume in patchPodChanges that those fields are the only ones that can change. This PR does the job, but it extends the existing code, even though part of it may not be needed anymore.

In order to make the code-base maintainable, could you consider changing AllocatePodMac to it does not mutate the pod? e.g. it could return a map of interfaces and their assigned MACs. We can generate the jsonPatch based on this information alone. That would limit the function interface to minimum, make it possible to test it and judge the behavior based on the function signature alone.

**Review by phoracek — CHANGES_REQUESTED** [review]

File: pkg/webhook/pod/pod.go
Comment (RamLavi) [reply pkg/webhook/pod/pod.go:72]: Can you elaborate exactly what you suggest to change? Do you mean only on pod's handler or also vm handler? 

In pods handler we only change the annotation, but in virtualmachine handler we add the interface's macAddress and also add finalizer. to common denominator is of the last 2 is the vm spec itself. Alternatively I can return a struct of all the changes (interfaces & finalizers) but I don't think it would make it more maintainable.
That's why instead of returning only the changes, and constructing the patches from them, I preferred in `patchPodChanges` and `patchVMChanges` to return the copy of the changed vm/pod and then extract exactly the needed changes from it.

Comment (phoracek) [reply pkg/webhook/pod/pod.go:72]: I would not even return the changes, just which MAC should be allocated to each VM. This way you'd be able to test the reading of the VM, figuring out whether we handle the allocation correctly.

The next function will get this map and apply it on a specific VM - return patches. I'm not sure I follow why'd you return finalizers. If MACs are allocated, you need a finalizer. If not, you don't need any. This can be decided by the second function.

I will not fight for this. Although I believe that having smaller functions with limited signatures would make the maintainability and testability easier.

Comment (RamLavi) [reply pkg/webhook/pod/pod.go:72]: I see what you mean, but changing the logic in the way you suggest needs more careful examination in order to unharm the current logic. Can we further discuss and do the change you suggest in another PR?

Comment (phoracek) [reply pkg/webhook/pod/pod.go:72]: Fair enough

**Review by phoracek — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
